### DISCUSSION TITLE
Run PocketBase Embedded in Caddy

### DISCUSSION THREAD

**Discussion by mohammed90** [pocketbase/pocketbase#5997] [Show and tell]
I noticed PocketBase can be used as a library, so, naturally, I wondered if it'd be possible to make it part of a @caddyserver plugin. With a little bit of wiggling, it works! This repository contains 3 Caddy modules:
- `pocketbase` app, which is a global instance of PocketBase and allows the user to configure some parts of it (listener -- if needed to listen outside localhost, data dir, and origins)
- `pocketbase` HTTP handler, which finds the globally configured PocketBase app and pass incoming requests directly to it. If the app isn't explicitly configured, reasonable defaults are assumed.
- Translate the `superuser` CLI commands into endpoints on Caddy admin interface, allowing the user to create the superuser by calling Caddy admin endpoint. This means it can benefit from Caddy secure remote administration and ACL of the admin interface.

Repository: https://github.com/mohammed90/caddy-pocketbase
Caddy docs page:  https://caddyserver.com/docs/json/apps/pocketbase/

**Comment by nicosql** [comment]
neat but I wouldn't recommend doing this in prod, why hamper your flexibility - am I missing something?

**Reply by mohammed90** [reply]
Hamper your flexibility how exactly? You need to deploy it behind reverse-proxy anyways

**Comment by dschissler** [comment]
For this level of deep integration I don't think that you should use `pocketbase.go`, aka `pocketbase.NewWithConfig`.  So not only would you be re-implementing / copy-pasting at lot of `examples/base/main.go` but also `./pocketbase.go` as well.  The issue is that `pocketbase.go` has a `rootCmd` on it, but what does that even mean for a caddy embedded app to have a Cobra command ?.

So you would have to setup the signal catching again.  Probably maybe though Caddy exposes that as an event that you would hook into instead of PB catching it (since it's just a guest module).

In my own projects where I've heavily modified Pocketbase I'm still using `pocketbase.go` because I figure that it's easier to hack and patch over any assumptions than it is to maintain another version of boring things like signal handle, command line args for the most essential things, etc.  But even then it's inconsistently abstracted because some core things remain, while other "*nice to haves*" like migrations and JS VM aren't.  I mean that Pocketbase wrapping could use some more well thought out wrappers and wizards than what it current does.

Perhaps you could could focus on stripping Pocketbase down to its `core.App` for Caddy in order to be able to suggest a direction for the developer (who shall not be named, shamed, begged, threatened, hollered, tagged or otherwise)  to refactor some of that.  It could benefit a long of scenarios - your more serious / extreme integration more than most.

**Reply by mohammed90** [reply]
Like most of the modules I push out, they're itches to scratch, feasibility check, and may develop over time. Some of your points are things I've thought about, so let me get to them.

> The issue is that `pocketbase.go` has a `rootCmd` on it, but what does that even mean for a caddy embedded app to have a Cobra command ?.

Caddy creates its own Cobra CMD that are detached from the config. PocketBase's commands are `serve` and `superuser`. I don't care for `serve` because Caddy handles that (and I manage PocketBase's listener internally), but I mapped the `superuser` commands to endpoints on Caddy's admin interface. For each command under `superuser`, there's an endpoint and HTTP method it maps to on the admin interface.

> So not only would you be re-implementing / copy-pasting at lot of `examples/base/main.go` but also `./pocketbase.go` as well

I did go through some of those files to understand what I need to handle. The linked example shows how to plug plugins as CLIs and JS, neither of which are showstopper. The main function of interacting with auth and collection functions are usable. For the signals, Caddy handles them and notifies my plugin for action. I call `Shutdown` on PocketBase's embedded HTTP server, and I trigger `OnTerminate` to call the `ResetBootstrapState` func.

I have feedback and almost wrote them in my post, but I want to crystalize them better before posting them. One of them is the hooking a custom logger. Caddy uses zap, which can be wrapped as a `slog.Handler`, but there's no way I can pass that to PocketBase. I can only receive your `slog.Logger`. The gain of wanting to inject Caddy's logger is because we offer pluggable encoders and writing destinations, which I couldn't pass to PocketBase. Luckily, Caddy offers control over log sink output, which means the user can still capture PocketBase's logs via the log sink config and pass it to a destination, but they still cannot control the format. (I realize now that I've braindumped the feedback already 😂 Sorry).

**Reply by dschissler** [reply]
> I can only receive your slog.Logger

Perhaps it was just your particular use of language, but I want to be clear that I'm not on the Pocketbase team.  Sometimes I write with commanding authority, but that is merely a coincidence.

Oh, so you're proxying to pocketbase serve.

`xcaddy build --with github.com/mohammed90/caddy-pocketbase`

Thanks for posting this.  I didn't know that Caddy provided this infrastructure.  This could also be a good way for me to create my own build that is to be used exclusively as an API and as a JSVM framework with several libraries available in the VM (by registering them).  This might be the easiest way to get a custom build going, and then it could also use serious Caddy features as well.

> I mapped the superuser commands to endpoints on Caddy's admin interface.

That is very cool.  How does Caddy secure these admin endpoints?

Maybe you could make the admin endpoint prefix customizable via that Caddy JSON setting?

> Caddy creates its own Cobra CMD that are detached from the config.

I suppose that if Caddy already depends upon Cobra then there isn't much of a reason to need to shave that part off of the build.

Although, my feeling after looking at the code for 10 minutes is that you're probably 80% there to replacing `pocketbase.PocketBase` with `core.App`.  I bet that you could accomplish that in one session.

**Reply by dschissler** [reply]
What you need is `func NewBaseApp(config BaseAppConfig) *BaseApp {` instead of `pocketbase.PocketBase`.  This is where the `Bootstrap` that you're using is implemented.  Then store it as a `core.App` (interface) as Pocketbase does internally.  Most of these `BaseAppConfig` settings have defaults that trigger by a zero value.  It's basically not any more work than what you're already doing, and it opens up the settings for more direct manipulation.

```golang
type BaseAppConfig struct {
	DBConnect        DBConnectFunc
	DataDir          string
	EncryptionEnv    string
	QueryTimeout     time.Duration
	DataMaxOpenConns int
	DataMaxIdleConns int
	AuxMaxOpenConns  int
	AuxMaxIdleConns  int
	IsDev            bool
}

func NewBaseApp(config BaseAppConfig) *BaseApp {
	app := &BaseApp{
		settings:            newDefaultSettings(),
		store:               store.New[any](nil),
		cron:                cron.New(),
		subscriptionsBroker: subscriptions.NewBroker(),
		config:              &config,
	}

	// apply config defaults
	if app.config.DBConnect == nil {
		app.config.DBConnect = DefaultDBConnect
	}
	if app.config.DataMaxOpenConns <= 0 {
		app.config.DataMaxOpenConns = DefaultDataMaxOpenConns
	}
	if app.config.DataMaxIdleConns <= 0 {
		app.config.DataMaxIdleConns = DefaultDataMaxIdleConns
	}
	if app.config.AuxMaxOpenConns <= 0 {
		app.config.AuxMaxOpenConns = DefaultAuxMaxOpenConns
	}
	if app.config.AuxMaxIdleConns <= 0 {
		app.config.AuxMaxIdleConns = DefaultAuxMaxIdleConns
	}
	if app.config.QueryTimeout <= 0 {
		app.config.QueryTimeout = DefaultQueryTimeout
	}

	app.initHooks()
	app.registerBaseHooks()

	return app
}
```
